src/Npgsql/NpgsqlConnection.cs

```diff
@@ -1374,21 +1374,48 @@ public bool Wait(int timeout)
         /// arrives, and exits immediately. The asynchronous message is delivered via the normal events
         /// (<see cref="Notification"/>, <see cref="Notice"/>).
         /// </summary>
+        /// <param name="timeout">
+        /// The time-out value, in milliseconds.
+        /// The default value is 0, which indicates an infinite time-out period.
+        /// Specifying -1 also indicates an infinite time-out period.
+        /// </param>
         /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="CancellationToken.None"/>.</param>
         [PublicAPI]
-        public Task WaitAsync(CancellationToken cancellationToken = default)
+        public Task<bool> WaitAsync(int timeout, CancellationToken cancellationToken = default)
         {
             if (cancellationToken.IsCancellationRequested)
-                return Task.FromCanceled(cancellationToken);
+                return Task.FromCanceled<bool>(cancellationToken);
             if (Settings.Multiplexing)
                 throw new NotSupportedException($"{nameof(Wait)} isn't supported in multiplexing mode");
 
             CheckReady();
 
             Log.Debug("Starting to wait asynchronously...", Connector!.Id);
-            return Connector!.WaitAsync(cancellationToken);
+            using (NoSynchronizationContextScope.Enter())
+                return Connector!.Wait(true, timeout, cancellationToken);
         }
 
+        /// <summary>
+        /// Waits asynchronously until an asynchronous PostgreSQL messages (e.g. a notification)
+        /// arrives, and exits immediately. The asynchronous message is delivered via the normal events
+        /// (<see cref="Notification"/>, <see cref="Notice"/>).
+        /// </summary>
+        /// <param name="timeout">
+        /// The time-out value as <see cref="TimeSpan"/>
+        /// </param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="CancellationToken.None"/>.</param>
+        [PublicAPI]
+        public Task<bool> WaitAsync(TimeSpan timeout, CancellationToken cancellationToken = default) => WaitAsync((int)timeout.TotalMilliseconds, cancellationToken);
+
+        /// <summary>
+        /// Waits asynchronously until an asynchronous PostgreSQL messages (e.g. a notification)
+        /// arrives, and exits immediately. The asynchronous message is delivered via the normal events
+        /// (<see cref="Notification"/>, <see cref="Notice"/>).
+        /// </summary>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="CancellationToken.None"/>.</param>
+        [PublicAPI]
+        public Task<bool> WaitAsync(CancellationToken cancellationToken = default) => WaitAsync(0, cancellationToken);
```

Comment:
Should this just return Task (non-generic) as before? The return value should mean whether a timeout occurred, but there's no timeout parameter here.

On that note, let's document what the return value actually means in the xml docs.

BTW If we change the return type this also becomes a breaking change.

Reply:
> Should this just return Task (non-generic) as before? The return value should mean whether a timeout occurred, but there's no timeout parameter here.

Yes, it should - my copy-pasting skills have failed me here 😅